import pysbd
from nltk.tokenize.punkt import PunktSentenceTokenizer
from corpus2JSON import validate_segmentation

examples = [". 3D reconstruction of the EC. (A) 3D reconstruction of the yeast MtRNAP EC generated from unstained, frozen particles. The atomic structure of human EC (PDB ID: 4BOC (24)) is fitted into the 3D reconstruction (correlation coefficient 0.74). (B) As in (A), but for a 3D reconstruction generated from negatively stained particles. The atomic structure of human EC (PDB ID: 4BOC (24)) is fitted into the 3D reconstruction (correlation coefficient 0.84). (C) As in (B), but for ECs prepared with an MtRNAP from which the first N-terminal 243 residues were deleted (spRNAPIV EC). The atomic structure of human EC (PDB ID: 4BOC (24)) is fitted into the 3D reconstruction (correlation coefficient 0.92). (D) Left: schematic representation of EC DNA scaffold used in (24). Template and non-template strands are dyed in cyan and yellow, respectively. Cyan: 28 nts long non-template strand with nucleotides not visible in X-ray structure indicated by dashed squares; Orange-yellow: 28 nts long template strand with nucleotide not visible in X-ray structure indicated by dashed square; Red: 14nts long RNA. Underlined nucleotides are ones that form a transcription bubble. Right: docking of the human MtRNAPEC (PDB ID: 4BOC) in the cryo-EM map of yeast MtRNAPEC, obtained in this study. Template and non-template strands are dyed in cyan and yellow, respectively. Upstream (5 NT) and downstream (5 T) DNA termini are emphasized with spheres colored in yellow and cyan respectively. (E) Left: schematic representation of EC DNA scaffold used in this study, aligned with the DNA presented in (A), using the upstream edge of transcription bubble as a reference, alignment point. 21 nts long template and non-template strands are dyed in orange (yellow) and cyan, respectively, while 10 nts long RNA in red. Underlined nucleotides are ones that form a transcription bubble. Right: docking of the human MtRNAPEC (PDB ID: 4BOC) in the cryo-EM map, obtained in this study. Template and non-template strands are dyed in cyan and yellow, respectively. Positions of the upstream (5 NT) and downstream (5 T) DNA termini, shown with spheres colored in yellow and cyan respectively, are adjusted according to the alignment. (F) Difference map in which the 3D reconstruction of negatively stained EC was subtracted from the 3D reconstruction obtained from EC particles in which the downstream DNA end was labeled with streptavidin ('downstream streptavidin EC'). The scale bar indicates a difference in positions of expected, theoretical (blue spheres) and experimental (streptavidin) downstream end of DNA (5 terminus of NT strand) (G) Difference map in which the 3D reconstruction of negatively stained spRNAPIV EC was subtracted from the 3D reconstruction obtained from spRNAPIV EC particles in which the downstream DNA end was labeled with streptavidin ('downstream streptavidin spRNAPIV EC') (H) As in (F), but this time the subtraction is performed between the 3D reconstruction of EC and a 3D reconstruction of EC in which the upstream end of the DNA is streptavidin-labeled ('upstream streptavidin EC'). Figure 6. Displacement of the upstream end of DNA during the IC-EC transition. (A) Difference map (gray volume) generated by subtracting the IC from the upstream streptavidin IC with the atomic structure of streptavidin (gray) placed on the large density difference. The docked atomic structure of human MtRNAPEC (PDB ID: 4BOC, DNA not shown) and the DNA of aligned T7RNAPIC (3E2E) are shown and labeled as in Figure 5. (B) As in (A) but the subtraction is between EC and 'upstream streptavidin EC'. (C) Cleavage patterns of Fe-BABE molecules conjugated to the mtRNAP suggest conformational differences between the N-terminal domains of the IC and EC. Each panel shows the cleavages on either the NT-or T-strands, as indicated, in reactions with mtRNAP conjugated with Fe-BABE at the indicated residue positions. 'G/A' = markers used to map the cleavage sites with respect to the +1 transcription start site; '\u2212' and '+' refer, respectively, to absence or addition of ascrobate/peroxide to activate cleavage (low levels of cutting are seen even in the absence of ascorbate/peroxide). NTPs allowing RNA extension to 0-13 nt, as indicated, were added to the reactions. Each panel also shows the MtRNAP structure in surface representation with \u2212, + and uncharged regions in red, blue, and white, respectively. The DNA is in stick representation with the T-strand in green and the NT-strand in cyan. The position of the residue(s) to which Fe-BABE was conjugated is labeled and highlighted in yellow, and the centers of the sites of cleavage by those conjugates are indicated by showing the corresponding DNA nucleotides in space-filling representation (T-strand cleavage by conjugates at residues 780, 1025, 1074 and 1078 is not shown as it was very low, likely reflecting the fact that the proximal part of the T-strand is buried in the template binding cleft and inaccessible). Values were determined by measuring the rate of dinucleotide synthesis in reactions with ATP only, or the rates of incorporation of NTPs into all transcripts ('Total') or only runoff transcript in reactions with all four NTPs (three NTPs at constant concentration and ATP varied). Deletion of the MTF1 C-terminus is seen to markedly increase K ATP for initiation on duplex, but not bubble, templates.",
"The educational project \"Molecular Imaging\"\" was conducted in summer 2008 for the first time. It was found that the \"passive\" part of the offer was well received in the last four semesters. A total of 539 students used the slides for revising the seminars. The \"active part\", which expects a lot of participation from the students, was viewed by 21 students from the 2 nd (n=11), 3 rd (n=8) and 5 th clinical semester (n=2) in five POL groups of 4-5 students each. Except for two students all participants have completed the relevant course. Both drop-outs indicated that the course was too effort-intensive and one student complained that the technical conditions are inadequate. The technical difficulties consisted of initial problems with the Skype conference, causing the start of the first POLevent to be delayed by 30 minutes. Amongst the other 19 participants, the seminars became a very active forum in which the students became increasingly independent. This was reflected, among other things, in the activity of the tutor who only had to intervene in one seminar on myocardial scintigraphy when the group wanted to place its main focus on coronary angiography rather than put nuclear medical diagnostics. Otherwise, the tutor essentially only had to carry out moderating duties. Working through the cases on average took 6 to 7 weeks and in respect to the intended learning goal, the factual results were positive. Due to the (so far) relatively small number of participants, no valid comparison between the \"classic face-to-face POL seminars\" can be made. From the perspective of the tutors it can be said, however, that both the results of the students' work and the subsequent discussion were at a very high level. The virtual POL seminar had the character of a teaching project, so far with only relatively few participants. Accordingly, our evaluation interest was not only quantitative but also qualitative. The starting point of this approach was the social scientific criticism of the quantitative survey with questionnaires, which allow mathematisation of the results presentation and reproducibility but with results that often remain superficial due to the suppression of factors or that may produce a false picture [13]. In the context of a complex teaching-learning situation a of lack of proximity to the problem and a lack of practical relevance can be the result. With a qualitative survey, however, we chose an open results approach, which highlights the teaching-learning situation in its inner argumentative and practical structure. This is important because the views of the relevant stakeholders will be reflected systematically. Against this background, the evaluation of the course was carried out both in a feedback session at the end of the course and via a brief online questionnaire (moodle), in which students graded four aspects of the course (from 1 \"very good\" to 6 \"bad fail\"). 17 of the 21 students evaluated the event (see Figure 1) and gave mostly positive assessments. It is striking that the technical implementation of the course, especially the first POL groups, received the worst assessment which can also be ascribed to initial problems of the tutors. In the feedback session all participants stated that they benefited from the seminar. A virtual seminar was uncharted territory for most participants and this is reflected in statements that the format of the seminar takes some getting used to. There was also consensus that a restriction of communication to the linguistic level is difficult because the nonverbal element is important, particularly in group work. It is therefore not surprising that four of the five small groups met in face-to-face during the working stage to discuss the case. From the perspective of teachers and the clinic too, the seminar was instructive, even labour intensive compared to the classical \"chalk and talk\" approach, with an average of 11 hours (6 hours in total in the online phase, 3 hours for evaluating the results and feedback to students and 2 hours for tutoring support).",
". Here, we show that acute exposure to 10 \u03bc M ouabain, which is the concentration we use for acute treatments, has a negligible eff ect on intracellular sodium concentration ( Fig. 2e ). As the on-rate of ouabain exceeds the off -rate 18, eff ective concentrations of ouabain can be considerably reduced when longer exposure times are used. A volume of 10 nM ouabain was added to the culture medium during the last 24 h of culture. A 24 h exposure of rat renal cells to 10 nM ouabain had no eff ect on intracellular sodium concentration ( Fig. 2f ). In the proliferation protocol ( Fig. 3d,e ), we observed signifi cant eff ects using 1 nM ouabain. Intracellular sodium measurements . Primary cultures of rat proximal tubule cells were prepared as described previously 10,39,40. Th e cells were loaded with 10 \u03bc M SBFI-AM ( Invitrogen ) for 45 min at room temperature, followed by washing and 30 min de-esterifi cation in normal media with 1 mM Probenecid ( Invitrogen ) at 37 \u00b0 C. Sodium measurements were taken at 35 \u00b0 C in a closed chamber ( FCS3 , Bioptechs ) with a cooled CCD camera ( ORCA-ERG , Hamamatsu ) mounted on an upright microscope ( Axioskop 2 FS , Zeiss ) with a Zeiss Fluar 40 \u00d7 1.3 NA oil objective. Excitation at 340 and 380 nm was performed with a monochromator ( Polychrome IV , TILL Photonics ). Data acquisition and analysis was performed using MetaFluor ( Molecular Devices ). Cells were superfused with Krebs-Ringer with 1 mM probenecid. Calibration was performed using perfusion buff er withExpression of Wt1 and Pax2 genes . Total RNA was extracted from a single embryonic kidney using the Aurum Total RNA Mini Kit ( Bio-Rad ) according to the manufacturer ' s recommendations. Relative real-time reverse transcription -PCR was performed according to the manufacturer ' s recommendations for iScript One-Step RT -PCR Kit with SYBR Green ( Bio-Rad ). Quantifi cation was performed in quadruplet for each sample. Commercial 18S Classic II primers ( Ambion ) were used to control genomic DNA contamination, to normalize RNA quality and starting quantity, as well as random tube-to-tube variation in reverse transcription (RT) and PCR reactions. Serial dilution of reference RNA sample as template was used to generate standard curve for each real-time RT -PCR experiment. Th e standard curve was constructed by plotting the log of the starting quantity of total RNA against the C t value. Th e starting quantity of Wt1 and Pax2 mRNA in each test sample was estimated using iQ iCycler soft ware ( Bio-Rad ) using values from the standard curve and normalized using 18S RNA as reference. Values are presented as % of control (10 % FBS in Figure 4a,b , 0.2 % FBS in Figure 4c,d ). Rat Wt1 mRNA structure 41 (accession number X69716) was used to select Wt1-specifi c primers: upper, 5 \u2032 -TGCCACACCCCTACCGACAGTT-3 \u2032 ; lower, 5 \u2032 -CTTCAAGG-TAGCTCCGAGGTTCATC-3 \u2032 . Th e position of rat Wt1 primers corresponds to a common part of all Wt1 mRNA splice variants. Th e rat Pax2 mRNA structure 42 (accession number XM_239083) was used to select Pax2-specifi c primers: upper, 5 \u2032 -TACACTGATCCTGCCCACATTAGA-3 \u2032 ; lower, 5 \u2032 -GGATAGGAAGGAC GCTCAAAGACT-3 \u2032 . Th e rat Pax2 PCR fragment corresponds to human Pax2 mRNA splice variants A and E (accession numbers NM_003987 and NM_003990, respectively).",
". Initiation events occur throughout the subtelomere in the human 5p segment in three cell lines, and replication forks progress through the segment in both directions. SMARD analysis of three different cell lines indicated that the 5p telomere is replicated primarily by forks progressing from the subtelomere to the telomere. Alignments of replicated molecules fully labeled with both IdU (red) and CldU (green) are shown. A map of the 5p locus is depicted above each alignment, with the positions of the FISH probes (blue bars below) used for identifying and orienting the molecules indicated. Vertical orange lines indicate the positions of the ends of the subtelomeric FISH signals used to align the molecules. The boundary between the subtelomere and telomere is delineated by a vertical blue line. Yellow arrows mark sites of transition from IdU incorporation to CldU incorporation and indicate the direction of fork progression at the moment of transition during the replication of the molecule. Replication profiles, histograms of the percentage of molecules containing IdU per 5-kb interval along the segment, are shown below each alignment. Initiation events (red tracts surrounded by green) occur at multiple locations. The origins appear to be clustered around a 40-90-kb region centered 65-115-kb from the telomere, seen as a peak in the replication profiles. Aside from a leftward broadening of the initiation zone in IMR-90 cells, there was not much variation in the basic features of the replication program of the 5p telomere in the two ES cell lines (H1 and H9) and the primary fibroblasts (IMR-90)."]

def test_sentence_segmentation(examples):
    seg = pysbd.Segmenter(language="en", clean=False, char_span=True)
    for example in examples:
        # pySBD
        the_sentences = seg.segment(example)
        sentences = []
        for the_span in the_sentences:
            span = {}
            span["start"] = the_span.start
            span["end"] = the_span.end
            sentences.append(span)
        if validate_segmentation(sentences):
            print("\n --> segmentation is valid")
        else:
            print("\n --> segmentation is not valid")
        print("nb. sentences:", str(len(sentences)))
        for span in sentences:    
            print(span["start"], span["end"])
            #print(example[span["start"]:span["end"]])
        # NLTK for comparison         
        sentences = []
        for start, end in PunktSentenceTokenizer().span_tokenize(example):
            span = {}
            span["start"] = start
            span["end"] = end
            sentences.append(span)
        if validate_segmentation(sentences):
            print("\n --> segmentation is valid")
        else:
            print("\n --> segmentation is not valid")
        print("nb. sentences:", str(len(sentences)))
        for span in sentences:
            print(span["start"], span["end"])
            #print(example[span["start"]:span["end"]])

if __name__ == "__main__":
    test_sentence_segmentation(examples)
